runtime/binding-ws/src/main/java/io/aklivity/zilla/runtime/binding/ws/internal/stream/WsServerFactory.java

```diff
@@ -215,25 +209,58 @@ else if (key != null &&
                     if (newRoute != null && (route == null || newRoute.order < route.order))
                     {
                         route = newRoute;
-                        protocol = newProtocol;
+                        websocketProtocol = newProtocol;
                     }
                 }
 
-                if (route != null)
+                if (route != null && WEBSOCKET_VERSION_13.equals(version))
                 {
-                    newStream = new WsServer(
-                        sender,
-                        originId,
-                        routedId,
-                        initialId,
-                        route.id,
-                        key,
-                        protocol,
-                        scheme,
-                        authority,
-                        path)::onNetMessage;
+                    final String key = headers.get("sec-websocket-key");
+                    if (key != null &&
+                        WEBSOCKET_UPGRADE.equalsIgnoreCase(upgrade))
+                    {
+                        // HTTP1.1 handshake
+                        newStream = new WsServer(
+                            sender,
+                            originId,
+                            routedId,
+                            initialId,
+                            route.id,
+                            HttpVersion.HTTP_1_1,
+                            key,
+                            websocketProtocol,
+                            scheme,
+                            authority,
+                            path)::onNetMessage;
+                    }
+                    else if (WEBSOCKET_PROTOCOL.equalsIgnoreCase(protocol))
+                    {
+                        // HTTP2 handshake
+                        newStream = new WsServer(
+                            sender,
+                            originId,
+                            routedId,
+                            initialId,
+                            route.id,
+                            HttpVersion.HTTP_2,
+                            null,
+                            websocketProtocol,
+                            scheme,
+                            authority,
+                            path)::onNetMessage;
+                    }
                 }
             }
+
+        }
+        else
+        {
+            final long newReplyId = supplyReplyId.applyAsLong(initialId);
+            doHttpBegin(sender, originId, routedId, newReplyId, 0L, 0L, 0, traceId, authorization, affinity,
+                hs -> hs.item(h -> h.name(":status").value("400"))
+                        .item(h -> h.name("connection").value("close")));
+            doHttpEnd(sender, originId, routedId, newReplyId, traceId);
+            newStream = (t, b, o, l) -> {};
```

Comment:
Do we need to send a `400` response explicitly here?
By default, returning `null` for the stream will trigger a `404` not found.
If you want to override the status, you can send back a `RESET` with `HttpResetEx` including `headers` of `:status` `400`, then `http` binding will send back a `400` for you.